kv/bulk: TestDuplicateHandling failed

kv/bulk.TestDuplicateHandling [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/16008841?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/16008841?buildTab=artifacts#/) on release-23.2.7-rc @ [8a4e55e40110d752c158414bfce22846faeb1722](https://github.com/cockroachdb/cockroach/commits/8a4e55e40110d752c158414bfce22846faeb1722):


```
=== RUN   TestDuplicateHandling
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/9cd32714a61c51d9ec44283a098261bc/logTestDuplicateHandling2115183443
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:146: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    conditional_wrap.go:190: 
        pkg/kv/bulk_test/pkg/kv/bulk/sst_batcher_test.go:201: (TestDuplicateHandling.func10)
        	NOTICE: .ClusterSettings() called via implicit interface ApplicationLayerInterface;
        HINT: consider using .ApplicationLayer().ClusterSettings() instead.
    conditional_wrap.go:194: TIP: consider replacing the test server initialization from:
            ts, ... := serverutils.StartServer(t, ...)
            defer ts.Stopper().Stop(...)
        to:
            srv, ... := serverutils.StartServer(t, ...)
            defer srv.Stopper().Stop(...)
            ts := srv.ApplicationLayer()
        
        See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization
    sst_batcher_test.go:92: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/bulk_test/pkg/kv/bulk/sst_batcher_test.go:92
        	            				github.com/cockroachdb/cockroach/pkg/kv/bulk_test/pkg/kv/bulk/sst_batcher_test.go:211
        	Error:      	Not equal: 
        	            	expected: 20
        	            	actual  : 10
        	Test:       	TestDuplicateHandling
    panic.go:523: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/9cd32714a61c51d9ec44283a098261bc/logTestDuplicateHandling2115183443
--- FAIL: TestDuplicateHandling (6.25s)
=== RUN   TestDuplicateHandling/ingestAll_does_not_error_on_key_matches_at_different_timestamps
    testing.go:1490: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
    --- FAIL: TestDuplicateHandling/ingestAll_does_not_error_on_key_matches_at_different_timestamps (0.03s)
```

Parameters:
 - <code>TAGS=bazel,gss,deadlock</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDuplicateHandling.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-40190

--- a/pkg/kv/bulk/sst_batcher_test.go
+++ b/pkg/kv/bulk/sst_batcher_test.go
@@ -57,14 +57,14 @@ func TestDuplicateHandling(t *testing.T) {
 	s, _, kvDB := serverutils.StartServer(t, base.TestServerArgs{})
 	defer s.Stopper().Stop(ctx)
 
-	expectRevisionCount := func(startKey roachpb.Key, endKey roachpb.Key, count int) {
+	expectRevisionCount := func(startKey roachpb.Key, endKey roachpb.Key, count int, exportStartTime hlc.Timestamp) {
 		req := &kvpb.ExportRequest{
 			RequestHeader: kvpb.RequestHeader{
 				Key:    startKey,
 				EndKey: endKey,
 			},
 			MVCCFilter: kvpb.MVCCFilter_All,
-			StartTime:  hlc.Timestamp{},
+			StartTime:  exportStartTime,
 		}
 		header := kvpb.Header{Timestamp: s.Clock().Now()}
 		resp, err := kv.SendWrappedWith(ctx,
@@ -99,11 +99,12 @@ func TestDuplicateHandling(t *testing.T) {
 	type keyBuilder func(i int, ts int) storage.MVCCKey
 
 	type testCase struct {
-		name           string
-		skipDuplicates bool
-		ingestAll      bool
-		addKeys        func(*testing.T, *bulk.SSTBatcher, keyBuilder) storage.MVCCKey
-		expectedCount  int
+		name            string
+		skipDuplicates  bool
+		ingestAll       bool
+		addKeys         func(*testing.T, *bulk.SSTBatcher, keyBuilder) storage.MVCCKey
+		expectedCount   int
+		exportStartTime hlc.Timestamp
 	}
 	testCases := []testCase{
 		{
@@ -122,6 +123,9 @@ func TestDuplicateHandling(t *testing.T) {
 		{
 			name:      "ingestAll does not error on key-value matches at different timestamps",
 			ingestAll: true,
+			// Set the export startTime to ensure all revisions are read, or fail if
+			// the gc threshold has advance past the start time
+			exportStartTime: hlc.Timestamp{WallTime: int64(tsStart) - 1},
 			addKeys: func(t *testing.T, b *bulk.SSTBatcher, k keyBuilder) storage.MVCCKey {
 				for i := 0; i < keyCount; i++ {
 					require.NoError(t, b.AddMVCCKey(ctx, k(i+1, tsStart+1), value))
@@ -134,6 +138,9 @@ func TestDuplicateHandling(t *testing.T) {
 		{
 			name:      "ingestAll does not error on key matches at different timestamps",
 			ingestAll: true,
+			// Set the export startTime to ensure all revisions are read, or fail if
+			// the gc threshold has advance past the start time
+			exportStartTime: hlc.Timestamp{WallTime: int64(tsStart) - 1},
 			addKeys: func(t *testing.T, b *bulk.SSTBatcher, k keyBuilder) storage.MVCCKey {
 				for i := 0; i < keyCount; i++ {
 					require.NoError(t, b.AddMVCCKey(ctx, k(i+1, tsStart+1), value))
@@ -208,7 +215,7 @@ func TestDuplicateHandling(t *testing.T) {
 			endKey := tc.addKeys(t, b, k)
 			if tc.expectedCount > 0 {
 				require.NoError(t, b.Flush(ctx))
-				expectRevisionCount(k(0, tsStart).Key, endKey.Key, tc.expectedCount)
+				expectRevisionCount(k(0, tsStart).Key, endKey.Key, tc.expectedCount, tc.exportStartTime)
 			}
 		})
 	}

